### PR TITLE
[draft] 11.0 UI notifications for jobs

### PR DESCRIPTION
**Pull request by guewen**
This is a draft of the API for job notifications. (https://twitter.com/guewenb/status/1000818392591462400)

There is code duplicated with [web_notify](https://github.com/OCA/web/tree/11.0/web_notify) which can be traded against a new dependency. 

https://github.com/OCA/web/tree/11.0/web_notifyhttps://github.com/OCA/web/tree/11.0/web_notify

The following events are triggered:

* A new job is enqueued
* Job started
* Job finished
* Job failed

You can also send a simple notification with: `notify` or `notify_warn`:

```python
            from odoo.addons.queue_job.job import job, notify

            @api.multi
            @job(on_done=notify('Partner exported'))
            def export_partner(self):
                # ...
```

For more advanced messages, you can also create methods on the `queue.job` model and refer to them with a string. This is also better for inheritance. From these hook methods, you can call `QueueJob.notify()` or  `QueueJob.notify_warn()`.

    Example usage:

```python

        class QueueJob(models.Model):
            _inherit = 'queue.job'

            @api.multi
            def on_export_partner_done(self):
                self.ensure_one()
                model = self.model_name
                partner = self.env[model].browse(self.record_ids)
                message = _('Partner %s exported') % partner.name
                self.notify(message=message)

        class ResPartner(models.Model):
            _inherit = 'res.partner'

            @api.multi
            @job(on_done='on_export_partner_done')
            def export_partner(self):
                # ...
```

### CODE REVIEW

File: queue_job/job.py [hunk 1]
```diff
@@ -620,14 +667,67 @@ def retryable_example():
...
+
+            from odoo.addons.queue_job.job import job, notify
+
+            @api.multi
+            @job(on_done=notify(message='Partner exported'))
+            def export_partner(self):
+                # ...
+
+    For more advanced messages, you can also create methods on the
+    ``queue.job`` model and refer to them with a string. From these
```
Comment (guewen) [inline queue_job/job.py:693]: note to myself: better inheritance than the first way

**Comment by guewen** [comment]
Looking for feedbacks on the notify API

**Comment by sbidoul** [comment]
@guewen I had not had time to look in  details yet. 

My first impression is that on_done notification is as easily (or more easily) implemented with an explicit call to notify (using web_notify) in the job function itself. The callback looks a bit weird to me where it can be implemented inline in the job function.

Same for job start. Same for on enqueued. 

on_error could benefit from some support of the queue framework. 

(it's just a first impression, certainly not a definitive opinion)

**Comment by guewen** [comment]
Thanks for your feedback! I want to find the right design and there is a lot of place for discussion :) I work on this on my spare time, no time constraint, no stress.

When you say inline, is it what you mean?

```python
class SaleOrder(models.Model):
    _inherit = 'sale.order'

    @job
    @api.multi
    def render_report(self):
        # 1. call to notify (job started)
        self._render()
        # 2. call to notify (job done)
```

This is an option I considered, and it would look straightforward, but at this point we miss the "author" of the job. `self` is the current sale order, and we don't know anymore what the job was, hence who triggered the job. Well we could retrieve the job with some value in the context but it wouldn't look cleaner IMO.

That's how I came to the conclusion (not definitive though) that the notify should be handled outside of the execution of the job function, from the executor of the job. 

I evaluated the idea that the `on_done`, `on_started`, ... triggers could serve other purposes by the way even if that could be YAGNI.

The `notify` method in `queue.job` is meant to call directly the method provided by `web_notify`, it should become:

```python
@api.multi
def notify(self, message, title=None, sticky=False):
    self.user_id.notify_info(message, title=title, sticky=sticky)
```

**Comment by sbidoul** [comment]
Still, having to inherit queue_job to implement the notification call feels a bit awkward, compared to the inline call. In the use cases I met, we did not need access to the job itself when notifying. And I kind of remember a PR somewhere that added the job UUID in the context so it should be accessible in cases we need it.

**Comment by lmignon** [comment]
@guewen I've the same feeling than @sbidoul. The notification can be easily done into the job function  if needed. On reading the comments, the missing parts is a method to get the running job on base model. As mentioned by @sbidoul everything is in place to easily implement such method. https://github.com/OCA/queue/blob/11.0/queue_job/job.py#L437

```python
class BaseModel(...):
    _inherit = 'base'

    def _get_running_job(self):
        job_uuid = self.env.context.get('job_uuid')
        if not job_uuid:
            return None
        return Job.load(self.env, job_uuid)
```

I don't know if this method should return a 'Job'  or  a 'queue.job' instance.

**Comment by guewen** [comment]
Then maybe we should find an easier way to get the job (record) from the inline job function by adding a generic function in the `base` model:

```python
class BaseModel(models.Model):
    _inherit = 'base'

    def outer_job(self):
        # we could pass the id of the job in the context btw
        return self.env['queue.job'].search([('uuid', '=', self.env.context.get('job_uuid'))])


[...]

    @job
    @api.multi
    def render_report(self):
        self.outer_job().notify('hello')
        # work
        self.outer_job().notify('goodbye')
```

This is simple, flexible and would make other things easier (like if we want to see the number of retries of the job: `self.outer_job().retries`... et voilà). I really like this possibility and maybe we should do it in any way.

Now I'm thinking about an issue with this approach: as the notification is sent after the commit of the transaction, it really handles only when a job is done. In my code above, `hello` and `goodbye` would be received at the same time by the user. The `on start` notification is probably much less interesting, but it can still be good to have.

Regarding the `on create` notification, sure, it can be handled "manually" by the code delaying the job:

```python
def confirm(self):
    self.env.user.notify('report will start in background')
    self.with_delay().render_report()
```
But then you have to duplicate that everywhere you are calling the job.

**Comment by guewen** [comment]
@lmignon our messages crossed with the same proposition, good sign :smile:

**Comment by sbidoul** [comment]
> notification is sent after the commit of the transaction

Ah I had forgotten that. But then web_notify could gain a nice option to notify in a separated short lived transaction.

**Comment by lmignon** [comment]
@guewen  :smile: 
@sbidoul I was just writing the same proposal... as stated by @guewen : good sign

**Comment by sbidoul** [comment]
Also, is it correct that jobs run in the context of the user that created them by default? In which case it's the main reason to access the job when notifying that disappears.

**Comment by guewen** [comment]
@sbidoul Oh yes good point, `self.env.user` is the same thing than `self.outer_job().user_id` so `self.env.user.notify(...)` is enough.

**Comment by guewen** [comment]
Any proposition for failures and create? 

Other options to consider:
* have boolean `notify_create` and `notify_failure` in the `job` decorator, message being a generic one such as "{job description} failed", "{job description} created"
* add parameters in `with_delay()`, which means we would have to store the messages in the job (in the language of the user). We could store them in a json field. It would not allow dynamic messages like "Job X failed after an execution time of x.y seconds after the 3 retry."

**Comment by lmignon** [comment]
@guewen 

> *  have boolean notify_create and notify_failure in the job decorator, message being a generic one such as " failed", " created"

Would this `notify_failure` parameter be redundant with mail_thread? We already get the ability to be notified by mail on job failure. 

> * add parameters in with_delay(), which means we would have to store the messages in the job (in the language of the user). We could store them in a json field. It would not allow dynamic messages like "Job X failed after an execution time of x.y seconds after the 3 retry."

It seems more natural to me to request the notification of the job creation when delaying the method. In this way we could notify the creation is some cases but not in others... But to avoid to store a new info in the ```job``` instance  we can also achieve the same result with a ```notify_create``` parameter in the ```job``` decorator and a new one into the ```delay``` method to mute the notification in some specific cases. Nevertheless if we add a new parameter in the ```job``` decorator for a message, this parameter will  not be seen as translatable by the translation system. (If I'm not wrong you can't use ```_('my message')``` into a decorator).  That means that we have to manually add this message into the translation files.

In short, I don't really have a clear opinion between one or the other approach.

**Comment by yelizariev** [comment]
Hello @guewen, thank you for your efforts! Job events is something really needed. 

Personally, I'm interested in API part only and cannot review UI part as I don't have usage scenarios... 

As for API, I wonder, is it possible to add ``on_done`` callback in ``with_delay`` call? It would be useful. Though the same can be implemented via several methods decorated with ``@job(on_done=...)``

Another question. Is it easy to split this PR in two, where one adds new API only and another depends on it and adds UI?
